File: android/app/src/main/java/com/satyam/birthdaybuddy/WidgetDataBridgeModule.java
package com.satyam.birthdaybuddy;

import android.content.Context;
import android.content.SharedPreferences;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Promise;

public class WidgetDataBridgeModule extends ReactContextBaseJavaModule {
    private static final String PREFS_NAME = "BirthdayWidgetPrefs";
    private static final String KEY_WIDGET_DATA = "birthday_widget_data";

    private final ReactApplicationContext reactContext;

    public WidgetDataBridgeModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @Override
    public String getName() {
        return "WidgetDataBridgeModule";
    }

    @ReactMethod
    public void saveWidgetData(String data, Promise promise) {
        try {
            SharedPreferences prefs = reactContext.getSharedPreferences(
                PREFS_NAME, 
                Context.MODE_PRIVATE
            );
            SharedPreferences.Editor editor = prefs.edit();
            editor.putString(KEY_WIDGET_DATA, data);
            editor.apply();
            
            android.util.Log.d("WidgetDataBridge", "Data saved successfully: " + data);
            promise.resolve(true);
        } catch (Exception e) {
            android.util.Log.e("WidgetDataBridge", "Failed to save data", e);
            promise.reject("SAVE_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void getWidgetData(Promise promise) {
        try {
            SharedPreferences prefs = reactContext.getSharedPreferences(
                PREFS_NAME, 
                Context.MODE_PRIVATE
            );
            String data = prefs.getString(KEY_WIDGET_DATA, null);
            
            android.util.Log.d("WidgetDataBridge", "Data retrieved: " + (data != null ? data : "NULL"));
            promise.resolve(data);
        } catch (Exception e) {
            android.util.Log.e("WidgetDataBridge", "Failed to get data", e);
            promise.reject("GET_ERROR", e.getMessage());
        }
    }

    @ReactMethod
    public void clearWidgetData(Promise promise) {
        try {
            SharedPreferences prefs = reactContext.getSharedPreferences(
                PREFS_NAME, 
                Context.MODE_PRIVATE
            );
            SharedPreferences.Editor editor = prefs.edit();
            editor.remove(KEY_WIDGET_DATA);
            editor.apply();
            
            android.util.Log.d("WidgetDataBridge", "Data cleared");
            promise.resolve(true);
        } catch (Exception e) {
            android.util.Log.e("WidgetDataBridge", "Failed to clear data", e);
            promise.reject("CLEAR_ERROR", e.getMessage());
        }
    }
}

File: modules/widget-data-bridge/android/src/main/java/expo/modules/widgetdatabridge/WidgetDataBridgeModule.kt
package expo.modules.widgetdatabridge

import android.content.Context
import android.content.SharedPreferences
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class WidgetDataBridgeModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("WidgetDataBridge")

    Function("saveWidgetData") { data: String ->
      val context = appContext.reactContext ?: return@Function false
      val prefs = context.getSharedPreferences("BirthdayWidgetPrefs", Context.MODE_PRIVATE)
      prefs.edit().putString("birthday_widget_data", data).apply()
      return@Function true
    }

    Function("getWidgetData") {
      val context = appContext.reactContext ?: return@Function null
      val prefs = context.getSharedPreferences("BirthdayWidgetPrefs", Context.MODE_PRIVATE)
      return@Function prefs.getString("birthday_widget_data", null)
    }

    Function("clearWidgetData") {
      val context = appContext.reactContext ?: return@Function false
      val prefs = context.getSharedPreferences("BirthdayWidgetPrefs", Context.MODE_PRIVATE)
      prefs.edit().remove("birthday_widget_data").apply()
      return@Function true
    }
  }
}
